challenge_02/base_solution.py

```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
     One of the most popular part in NLP.
     If you understand self attention, you will understand the heard of transformers.
     """
-    def __init__(self):
+    def __init__(self, embedding_length: int) -> None:
         super().__init__()
-        self.W = None
+        self.e_length = embedding_length
+        self.TW = self._trainable_weights()
+
+    def _trainable_weights(self) -> Tuple[nn.Module, nn.Module, nn.Module]:
+        q = nn.Linear(self.e_length, self.e_length, bias=False)
+        k = nn.Linear(self.e_length, self.e_length, bias=False)
+        v = nn.Linear(self.e_length, self.e_length, bias=False)
+        return q, k, v
 
-    def forward(self, x):
-        q, k, v = self.W
-        # norm_attention = F.softmax(q*k)
-        # (norm_attention * v).sum(-1)
-        return
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Scaled Dot-Product Attention
+        q, k, v = self.TW
+        s = torch.matmul(q(x), torch.t(k(x)))
+        w = nn.functional.normalize(s)
+        w = F.softmax(w)
+        return torch.matmul(w, v(x))
 
 if __name__ == "__main__":
-    x = torch.randint(0, 500, (10, 50))  # tokenizer(text)
-    model = SelfAttention()
+    embdd_len = 50
+    x = torch.randint(0, 500, (10, embdd_len), dtype=torch.float32)  # tokenizer(text)
```

Comment:
spíš testuj jestli ti to funguje na [batch_size, tokens, embd_dim]

Reply:
to se mi v minulosti vymstilo že jsem testoval pro jeden example bez dimenze batche a potom to nefungovalo 😄